submissions/misshatsker/object-oriented-js/app.js

```diff
@@ -0,0 +1,188 @@
+    this.speed = speed;
+    this.x = x;
+    this.y = y;
+};
+
+Character.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const Enemy = function(...args) {
+    Character.call(this, ...args);
+};
+
+Enemy.prototype.update = function(dt) {
+    const shift = this.speed * dt;
+    this.x = this.x + shift;
+    if (this.x >= rightSide) {
+        this.x = -this.width;
+    }
+};
+
+Enemy.prototype.width = enemyWidth;
+Enemy.prototype.height = enemyHeight;
+
+Object.setPrototypeOf(Enemy.prototype, Character.prototype);
+
+const Player = function(...args) {
+    Character.call(this, ...args);
+};
+
+Player.prototype.width = playerWidth;
+Player.prototype.height = playerHeight;
+
+Player.prototype.update = function() {
+    if (checkPlayerOverlapWithEnemies(this)) {
+        gameOver(false);
+    }
+}
+
+Player.prototype.handleInput = function(action) {
+    switch (action) {
+        case 'left': {
+            if (this.x > leftSide) {
+                this.x -= this.speed;
+            }
+            break;
+        }
+        case 'right': {
+            if (this.x + this.width + this.speed < rightSide) {
+                this.x += this.speed;
+            }
+            break;
+        }
+        case 'down': {
+            if (this.y + this.height + this.speed < bottomSide) {
+                this.y += this.speed;
+            }
+            break;
+        }
+        case 'up': {
+            this.y -= this.speed;
+            if (this.y <= 0) {
+                gameOver(true);
+            }
+
+            break;
+        }
+    }
+}
+
+Player.prototype.setPosition = function(x, y) {
+    this.x = x;
+    this.y = y;
+}
+
+Object.setPrototypeOf(Player.prototype, Character.prototype);
+
+function getRandomNumberFromARange(min, max) {
+    return (Math.random() * (max - min)) + min;
+}
+
+function createRandomEnemy(row) {
+    const x = getRandomNumberFromARange(enemyMinX, enemyMaxX);
+    const y = initialGroundY + row * tileHeight;
+    const speed = getRandomNumberFromARange(enemyMinSpeed, enemyMaxSpeed);
+
+    return new Enemy(x, y, speed, enemySprite);
+}
+
+const allEnemies = [
+    createRandomEnemy(0),
+    createRandomEnemy(1), 
+    createRandomEnemy(2)
+];
+
+const player = new Player(playerInitialX, playerInitialY, playerSpeed, playerSprite);
+
+document.addEventListener('keyup', function(e) {
+    const allowedKeys = {
+        37: 'left',
+        38: 'up',
+        39: 'right',
+        40: 'down'
+    };
+
+    player.handleInput(allowedKeys[e.keyCode]);
+});
+
+function checkIsOverlap(l1, r1,  l2,  r2) {
+    const isRectangleOnLeftSideOfAnother = (l1.x > r2.x || l2.x > r1.x);
+    if (isRectangleOnLeftSideOfAnother) {
+        return false;
+    }
+
+    const isOneRectangleAboveAnother = (r1.y < l2.y || r2.y < l1.y);
+    if (isOneRectangleAboveAnother) {
+        return false;
+    }
+
+    return true;
+}
+
+function checkPlayerOverlapWithEnemies(player) {
```

Comment:
Why not a part of Enemy's or Player's behaviour?
We could build the entire game on global functions.
There is a reason for OOP to exist.

Reply:
I've update it, check it please